### ISSUE TITLE
Rotated files can lead to silently losing data

### ISSUE THREAD

**Issue by avleen** [elastic/logstash-forwarder#197]
This is a bug with a high risk of data loss.

When a file is rotated from under logstash-forwarder, the daemon keeps the file open until it has finished processing. In our case, this can sometimes be many hours.
During this time, logstash-forwarder has both the old file and the new file open, and it saves the state/offset for the file **using the same key**
Each time we cat the state file, we see a different value, corresponding to the positions of the old file and the new file:

```
    "source":"/var/log/httpd/info.log","offset":2089783802,"inode":926528
    "source":"/var/log/httpd/info.log","offset":926667352,"inode":926531
    "source":"/var/log/httpd/info.log","offset":2093587287,"inode":926528
    "source":"/var/log/httpd/info.log","offset":936770792,"inode":926531
    "source":"/var/log/httpd/info.log","offset":2206839913,"inode":926528
````

Depending on the which state is saved last, logstash-forwarder could end up re-reading existing log lines or skipping a large number of log lines, with no warning.
If a file is renamed, the daemon should state saving the state with the new filename.
If the file is deleted, the daemon should stop saving its state, as it wouldn't be able to recover it after a restart anyway.

**Comment by jordansissel** [comment]
Is this still an issue?

**Comment by jordansissel** [comment]
I *think* this was fixed in 0.4.0 where multiple file watchers were accidentally overwriting eachothers' state in the registrar.
